app/scripts/lib/derived-keys.js

```diff
@@ -0,0 +1,74 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
```

Comment:
I'd like to soon see an alternate implementation of this functionality in another language to sanity check it's correctness of what it appears to be doing.